app/common/auth/decorators.py

```diff
@@ -0,0 +1,37 @@
+import functools
+from typing import Callable, cast
+
+from flask import abort, redirect, request, session, url_for
+from flask.typing import ResponseReturnValue
+from flask_login import current_user
+
+from app.common.data.models import User
+
+
+def login_required[**P](
+    func: Callable[P, ResponseReturnValue],
+) -> Callable[P, ResponseReturnValue]:
+    @functools.wraps(func)
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> ResponseReturnValue:
+        if not current_user.is_authenticated:
+            session["next"] = request.full_path
+            return redirect(url_for("auth.request_a_link_to_sign_in"))
+
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def mhclg_login_required[**P](
+    func: Callable[P, ResponseReturnValue],
+) -> Callable[P, ResponseReturnValue]:
+    @functools.wraps(func)
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> ResponseReturnValue:
+        # This decorator is itself wrapped by `login_required`, so we know that `current_user` exists and is
+        # not an anonymous user (ie a user is definitely logged-in) if we get here.
+        if not cast(User, current_user).email.endswith("@communities.gov.uk"):
```

Comment:
Something about having this wrapper doing this fixed check independently of sign up which will also be checking this feels like it might lead to behavior thats hard to track down?

I think this only seems confusing because there aren't permissions/ scope to check (i.e "Access to grant", "Update grant", "Read grant") but once those are in we should be able to trust that for them to have those permissions they would have had good reason (i.e had the appropriate email domain for access)

I don't think this needs to change this PR as this is basically a single combined implementation of the authentication and permissions that don't otherwise exist! Just prompted starting to think it through

Reply:
Yep I think a good call out.

I'm fairly sure this will all come out in the wash when we start adding orgs and permissions properly, so confident to leave it in for now.

Fingers crossed I won't regret this arrogance.